[댓글 기능] 제이 미션 제출합니다.

JunHoPark93 opened the pull request:
미션이 쉽지 않네요.
요청은 dto를 정의해서 받았고 응답은 댓글만 응답 dto를 서비스 레이어에서 정의했습니다. (댓글시간 계산 필드 때문에)
나머지 user나 article 도 전부 다 dto를 정의해보려다 너무 복잡해지는것 같아서 그냥 domain으로 뷰에 내렸는데 괜찮을까요?
항상 감사드립니다!

MrKwon committed:
docs: Readme 업데이트

MrKwon committed:
feat: 게시글 생성, 목록 조회 기능 구현

MrKwon committed:
feat: 게시글 조회 구현

MrKwon committed:
feat: 공통 Head 분리

JunHoPark93 committed:
feat: html 중복제거

JunHoPark93 committed:
feat: 게시글 수정

JunHoPark93 committed:
feat: 게시글 삭제 구현

JunHoPark93 committed:
feat: devtools 적용

JunHoPark93 committed:
refactor: Js 분리

JunHoPark93 committed:
refactor: 컨트롤러 메서드 이름 수정, autowired 삭제

JunHoPark93 committed:
refactor: 불필요한 메서드 제거

JunHoPark93 committed:
feat: Article 제목 검증 추가

JunHoPark93 committed:
feat: error 페이지 추가

JunHoPark93 committed:
refactor: 피드백 반영

리다이렉트, 이름 잘못 수정된것 원복, header location 검증 추가

JunHoPark93 committed:
feat: Jpa repository 적용

JunHoPark93 committed:
Merge branch 'temp' into JunHoPark93

# Conflicts:
#	README.md
#	build.gradle
#	src/main/java/techcourse/myblog/domain/Article.java
#	src/main/java/techcourse/myblog/web/ArticleController.java
#	src/main/resources/application.properties
#	src/main/resources/templates/article.html
#	src/test/java/techcourse/myblog/web/ArticleControllerTests.java

JunHoPark93 committed:
docs: README 업데이트

JunHoPark93 committed:
feat: 회원 등록 기능

JunHoPark93 committed:
feat: 유저 조회 기능 구현

Squashed commit of the following:

commit 3026e5c7eb726b65b992756d9ec17a73bec140d4
Author: JunHoPark <pjh374968@gmail.com>
Date:   Tue Jul 16 17:54:01 2019 +0900

    feat: 유저 조회 기능 구현

commit c82f723cc643fd9aed7686d0231ed18cd35924b7
Author: JunHoPark <pjh374968@gmail.com>
Date:   Tue Jul 16 16:53:26 2019 +0900

    fix: Valid 애노테이션 추가

    Validation 에러 수정, @Valid가 없으면 검증이 발동되지 않는다

JunHoPark93 committed:
feat: 로그인 및 로그아웃 기능 구현

Squashed commit of the following:

commit 3af160545dc0fcdce403342652fd1c9497e60982
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 15:57:01 2019 +0900

    refactor: html 패키지화 원복

commit abaa172ae7507edcc0a8a21852291f5b216d967a
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 15:53:36 2019 +0900

    feat: 로그아웃 기능 구현

commit a523c371383247a05158711cdc7a1d1c6cd1c2ff
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 15:26:37 2019 +0900

    feat: 로그인 기능 구현

commit 010778b87f6995e0d38a29d15b3041529cb43b20
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 15:25:59 2019 +0900

    chore: JPA 실행 쿼리 보기 설정

commit fcf46a795cfdb3dcdd14a683d93dd1c1ea0817b7
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 14:48:54 2019 +0900

    feat: 세션 검증 Interceptor 구현

commit 261149e005613cc1fe6ab4ead0485842ff80780c
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 13:12:06 2019 +0900

    refactor: html 패키지 구조 변경

JunHoPark93 committed:
feat: 회원 수정 및 탈퇴기능 구현

Squashed commit of the following:

commit d36b93b9f56d5a631168ea59425740e3852b4c23
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 17:29:53 2019 +0900

    feat: 회원 탈퇴 기능 구현

commit 0a0adfd710c88ca96f77b0a14b71812556c0643d
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 17:07:59 2019 +0900

    feat: 회원 수정 기능 구현

JunHoPark93 committed:
refactor: html 중복제거 및 세션적용

JunHoPark93 committed:
fix: Interceptor에서 /users post요청은 통과

JunHoPark93 committed:
feat: Error 객체의 정의

Login 관련 에러 처리 객체

JunHoPark93 committed:
test: 세션관련 인수테스트 작성

JunHoPark93 committed:
test: 회원관련 인수테스트

JunHoPark93 committed:
refactor: 들여쓰기 및 사용하지 않는 import 제거

JunHoPark93 committed:
feat: 회원가입시 이메일 중복 오류 처리

JunHoPark93 committed:
fix: 회원 가입시 email 중복 테스트 수정

JunHoPark93 committed:
refactor: UserService 분리

JunHoPark93 committed:
feat: 회원 수정시 이름 형식 오류 처리

JunHoPark93 committed:
feat: 회원 가입시 유효하지 않은 입력 처리

JunHoPark93 committed:
refactor: Article 서비스 패키지 분리

JunHoPark93 committed:
fix: 유저 리스트에서 다른 유저 정보에 접근하는 오류 처리

JunHoPark93 committed:
feat: 데이터베이스 구축

* 프로덕션 환경은 mysql, 테스트 환경은 h2를 사용하였다.

JunHoPark93 committed:
refactor: 삼항 연산자로 변경

* login, logout을 구분하는 if-else문을 삼항 연산자로 변경

* article의 기본 그림을 구분하는 if-else문을 삼항 연산자로 변경

JunHoPark93 committed:
refactor: 삼항 연산자로 변경

JunHoPark93 committed:
fix: 게시글 id 접근하는 로직 수정

JunHoPark93 committed:
fix: 절대 경로로 변경

JunHoPark93 committed:
refactor: 삼항 연산자로 변경, url 변경

JunHoPark93 committed:
refactor: fragments 패키지로 분리

JunHoPark93 committed:
refactor: 피드백 반영

메서드 이름 수정
패키지 구조 전체적으로 변경 (root 기준으로 좀 더 깔끔해 지게 변경, 레이어드 아키텍쳐를 고려하여 depth 설정)
JPA 실행 쿼리 보여주기
JPA repository의 사용
Domain을 생성하는 곳 통일 (dto에서 domain 생성 vs 서비스에서 domain 생성 -> 서비스로 통일)

JunHoPark93 committed:
feat: 비밀번호 암호화 구현

BCrypt 라이브러리를 이용한다. 내 프로젝트에 맞게 wrapping 작업을 한다. 그 이유는 크게 두가지이다.
1. BCrypt 클래스의 이름과 내부 메서드 철자의 어려움
2. 프로젝트의 특성에 맞게 새로 클래스를 구현할 수 있음 (인터페이스로 분리하여 다른 암호화를 쉽게 Bean을 통해 주입받아 변경 가능)

마지막으로 비밀번호 검증은 Web Config 와 연관있다고 생각하여 interceptor를 구현한 클래스를 수정하여 Web 설정과 관련된 @Bean들의 집합 클래스로 변경

JunHoPark93 committed:
refactor: implements 삭제

JunHoPark93 committed:
feat: 비밀번호, 이메일 검증 로직 개선

JunHoPark93 committed:
feat: 검증 로직 추가

비밀번호, 비밀번호 확인 필드 검증 기능
회원 수정 dto 추가

JunHoPark93 committed:
fix: 없는 게시물 접근 처리

JunHoPark93 committed:
test: 테스트 파라미터 추가

JunHoPark93 committed:
fix: email validation 추가

JunHoPark93 committed:
test: cookie를 이용한 테스트

JunHoPark93 committed:
refactor: 피드백 반영

생성자 private, location 검증 수정

JunHoPark93 committed:
refactor: localport 제거

JunHoPark93 committed:
refactor: 패키지 이름 변경

JunHoPark93 committed:
refactor: 클래스 이름 변경

Impl 네이밍은 불필요하다. 조금 더 암호화 구체 클래스의 기능을 나타내는 이름으로 변경하였다.

JunHoPark93 committed:
Merge remote-tracking branch 'upstream/JunHoPark93' into JunHoPark93

JunHoPark93 committed:
feat: paging 적용

JunHoPark93 committed:
feat: article, user 관계 매핑

JunHoPark93 committed:
feat: article 작성자 매핑 기능 추가

JunHoPark93 committed:
feat: Comment 작성, 조회

JunHoPark93 committed:
feat: Comment 수정, 삭제

JunHoPark93 committed:
refactor: comment redirect

JunHoPark93 committed:
feat: Comment 시간 계산

JunHoPark93 committed:
fix: 타인이 게시글 삭제 막기

JunHoPark93 committed:
test: 삭제 테스트 수정 및 정리

interface 안쓰는것 삭제

woowahanCU reviewed (commented):
SaltEncryptTest는 SpringBootTest일 필요가 없지 않을까요

```java
class SaltEncryptTest {
    private EncryptHelper encryptHelper = new SaltEncrypt();
```

woowahanCU reviewed (commented):
AuthInterceptorTest는 모두 실패하고 있네요.

woowahanCU reviewed (commented):
GET /articles/dd HTTP/1.1" 400 7001

GetMapping("/articles/{articleId}" 요청시에 
불필요한 요청이 날아가네요.

woowahanCU reviewed (changes requested):
안녕하세요! 리뷰어 CU입니다.
짧은 시간내에 전반적으로 깔끔하게 잘 작성하셨습니다! 👍 
몇가지 피드백 남겨두었으니 확인해보시고 궁금한 점 있으면 DM 남겨주세요~ 고맙습니다.

woowahanCU commented:
추가적으로 질문하신 부분에 대해서 답드리자면, 
우선 요청과 응답의 시그니처는 맞춰주는것이 좋다고 생각합니다.
그리고 반드시 DTO를 작성해야 할 필요는 없습니다. 
Presentation Layer의 변경사항이 도메인 모델에 영향을 미치는 것을 줄이고자 작성해준다는 관점으로 생각하심이 좋을거 같네요.

JunHoPark93 committed:
feat: JPA auditing 적용

Article과 Comment는 User의 행위가 연관되어 있기 때문에 그 둘을 연관시켜 저장하는 auditing기능을 구현한다.

JunHoPark93 committed:
feat: Comment 응답 dto에 대한 Custom Mapper의 구현

custom Converter를 구현한다.

JunHoPark93 committed:
refactor: test 중복제거

JunHoPark93 committed:
feat: JPA auditing, custom Context holder 적용

수정자에 대한 업데이트가 이루어지도록 추가, 이 부분을 추가하려면 authenticated된 유저정보가 필요한데 security를 적용하지 않을 것이므로 static으로 유저정보들을 thread-safe하게 공유할 수 있는 contextholder를 직접 만들어본다.

JunHoPark93 committed:
fix: session holder 코드 원복

미션 기한을 맞추기 위해서 추후 기능으로 한다

JunHoPark93 committed:
refactor: Comment 검증 로직

User 체크를 조금 더 명시적이게 변경

JunHoPark93 commented:
리뷰 감사합니다!
지적해주신 부분들 반영해보았습니다.
## 1. JPA Auditing
보내주신 링크를 정독해 보았습니다. 링크안의 @CreatedBy를 쓰게 된다면 수정자도 auditing으로 구현할 수 있다는 것이 매우 의미있다고 생각했습니다. 그런데 @CreatedBy를 쓰려면 User 객체를 전역적으로 가지고 있다가 JPA 설정파일을 만들어 auditorAwareRef에 넣어줘야 했습니다. 
### 문제상황: 
저희 앱 자체가 context에 user정보를 가지고 있을 수가 없었습니다. 시큐리티를 썼다면 context에서 코드 한줄이면 현재 logged in 한 유저를 가져올 수 있었지만 지금은 쓸 필요가 없다고 생각해서 (써본적도없고) SessionContextHolder라는 thread-safe한 클래스를 만들어보려고 했습니다. 이 클래스는 제가만든 UserAuthentication 객체 (필드는 현재 한 클라이언트에서 부여받은 session id) 를 key로 user객체를 value로 가지고 있는 concurrentHashmap으로 구현했었습니다. 어떤 사용자가 로그인 시 앞단에서 검증을 통해 현재 서버의 sessionContext에 user객체가 존재한다면 (어디선가 로그인이 된 상태) 로그인을 막거나/그 사람을 튕기거나 둘 중에 하나를 구현하고 있었습니다. 이틀 동안 해본결과 오늘 아침 샤워하다가 초기 설계중 한 부분이 잘못되었다는 것을 뒤늦게 깨달았습니다. 이유는 Bean이 어떻게 관리되는지, thread가 어떻게 돌아가는지에 대한 명확한 이해가 없었기 때문에 로직이 어느정도 완성되었지만 치명적인 문제가 있었습니다. 
마감 시간이 중요하다고 생각해서 했던 작업들은 전부 날리고, @CreatedBy도 지우고 리뷰요청을 남깁니다. 이 부분은 미션이 끝나더라도 따로 작업을해서 꼭 완성해볼 예정입니다. 

## Dto 매핑 위임
ModelMapper를 활용해보라는 피드백을 받고 구현을 해보았습니다. CommentResponseMapper라는 매핑 클래스를 정의해보았습니다.

## 메서드 호출 의도
지적해주신 Comment의 검증부분이 말씀주신것처럼 의도가 잘 드러나지않은것 같았습니다. 조금 더 의도를 드러내기위해서 comment 객체에게 auther를 물어보는 형식으로 바꿔보았습니다.

## test context제거
SaltEncrypt 클래스는 서버를 띄울 필요가 없는 테스트인데 아무생각없이 붙였습니다.. 바로 지웠습니다.

## 기타
url 매핑 하는 컨트롤러에서 공통 된 부분을 위로 뺐습니다. 테스트의 중복코드를 조금 더 줄여보았습니다.

JunHoPark93 commented:
```code
GET /articles/dd HTTP/1.1" 400 7001

GetMapping("/articles/{articleId}" 요청시에
불필요한 요청이 날아가네요.
```

라고 말씀해주신 부분 확인해보았는데 article의 coverUrl의 이미지 url을 찾지 못하였을 때 발생되는것으로 보입니다. 

게시글을 대충 생성했을 때 coverUrl필드에 dd, df 그냥 이런 문자열을 저장한 이후 -> 프런트에서 image 태그에 그 url로 요청을보내서 resource를 가지고 오지 못하여 400대 에러가 발생합니다.

혹시 유효한 coverUrl인지 까지 Article객체를 생성할 때 검사를 하라는 것을 말씀하신건가용? 지금 생각나는 것은 article 객체를 저장할때 coverUrl을 restTemplate으로 찔러봐서 유효한 resource인지 검사하는 로직까지 넣는것을 의도하신건지 궁금합니다!

JunHoPark93 committed:
feat: CoverUrl resource의 존재 확인

head 요청으로 리소스의 유무를 확인하여 없다면 article 저장이 되지 않게 함

JunHoPark93 committed:
test: coverUrl 경로 수정

JunHoPark93 commented:
자답입니다! head 요청으로 resource 유뮤를 확인하여 없다면 게시글 저장이 되지 않게 하였습니다. 없는 url요청하는 불필요한 요청을 없앴습니다

woowahanCU reviewed (approved):
깔끔하게 잘 구현하셨네요! 
SaltEncrypt를  작성하여 Password를 암호화한 부분, AuthInterceptor와 SessionResolver를 적용한 부분, 그리고 Helper 클래스를 작성하여 테스트코드의 중복을 제거한 부분 등이 상당히 인상적이었어요! 💯
앞으로 남은 미션들에서도 많은 것을 학습하고 성장하길 바래요~ 
고맙습니다. 😀
